modules/broker-core/src/main/java/io/ballerina/messaging/broker/core/ObservableQueue.java

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright (c) 2018, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ *
+ * WSO2 Inc. licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.ballerina.messaging.broker.core;
+
+import io.ballerina.messaging.broker.eventing.EventSync;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import javax.transaction.xa.Xid;
+
+/**
+ * Represents an Queue which trigger events for the broker.
+ */
+public class ObservableQueue extends Queue {
+
+    private Queue queue;
+    private HashSet<Integer> messageLimits;
+    private EventSync eventSync;
+
+    ObservableQueue(Queue queue, EventSync eventSync, List<Integer> messageLimits) {
+        super(queue.getName(), queue.isDurable(), queue.isAutoDelete());
+        this.queue = queue;
+        this.eventSync = eventSync;
+        this.messageLimits = new HashSet<>(messageLimits);
```

Comment:
Shall we note why we are using a hash set instead of a concurrent set since this object is accessed concurrently?